<a href="https://colab.research.google.com/github/Manass20/Play-Store-App-Review-Analysis---Capstone-Project/blob/main/Manas_Ranjan_Play_Store_App_Review_Analysis_Capstone_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## <b> The Play Store apps data has enormous potential to drive app-making businesses to success. Actionable insights can be drawn for developers to work on and capture the Android market. </b>

## <b> Each app (row) has values for catergory, rating, size, and more. Another dataset contains customer reviews of the android apps.</b>

## <b> Explore and analyze the data to discover key factors responsible for app engagement and success. </b>

##**Introduction**

Play Store is a platform for browsing and downloading Android applications provided by Google. It is an official app store for certified Android applications developed using the Android Software Development Kit. It includes both free as well as paid apps.
 Our study used two datasets, one dataset consisted of information on play store apps, their ratings, and the basic information about the apps, and another was a user review dataset, which provided users' opinions of the apps.


# **FIRST STEP :-**




# **IMPORTING THE NECESSARY LIBRARIES**

In [56]:
import pandas as pd  # pandas library

import numpy as np   # numpy library

import matplotlib  

import matplotlib.pyplot as plt
%matplotlib inline

import seaborn as sns

## **IMPORTING THE DATASET'S INTO OUR COLLAB NOTEBOOK**

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


## **Below are the libraries that are used in order to perform EDA (Exploratory data analysis):**

*   Loading the data into the pandas data frame is certainly one of the most important steps in EDA, as we can see that the value from the data set is comma-separated

In [3]:
path='/content/drive/MyDrive/Colab Notebooks/AlmaBetter/PROJECTS/EDA/PLAY STORE/Play Store Data.csv'
store_df= pd.read_csv(path)
path_userRatings='/content/drive/MyDrive/Colab Notebooks/AlmaBetter/PROJECTS/EDA/PLAY STORE/User Reviews.csv'
ratings_df= pd.read_csv(path_userRatings)

## **SECOND STEP:-**

## **DATA EXPLORATION**

In [4]:
#Displaying the top 5 rows
store_df.head(5)

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  Reviews  Size     Installs  Type Price Content Rating  \
0     159   19M      10,000+  Free     0       Everyone   
1     967   14M     500,000+  Free     0       Everyone   
2   87510  8.7M   5,000,000+  Free     0       Everyone   
3  215644   25M  50,000,000+  Free     0           Teen   
4     967  2.8M     100,000+  Free     0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

In [5]:
#Displaying the bottom 10 rows
store_df.tail(10)

App             Category  \
10831                       payermonstationnement.fr  MAPS_AND_NAVIGATION   
10832                                       FR Tides              WEATHER   
10833                                    Chemin (fr)  BOOKS_AND_REFERENCE   
10834                                  FR Calculator               FAMILY   
10835                                       FR Forms             BUSINESS   
10836                               Sya9a Maroc - FR               FAMILY   
10837               Fr. Mike Schmitz Audio Teachings               FAMILY   
10838                         Parkinson Exercices FR              MEDICAL   
10839                  The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840  iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       Rating Reviews                Size     Installs  Type Price  \
10831     NaN      38                9.8M       5,000+  Free     0   
10832     3.8    1195                582k     100,000+  Free     0   
10833     4.8      44                619k       1,000+  Free     0   
10834     4.0       7                2.6M         500+  Free     0   
10835     NaN       0                9.6M          10+  Free     0   
10836     4.5      38                 53M       5,000+  Free     0   
10837     5.0       4                3.6M         100+  Free     0   
10838     NaN       3                9.5M       1,000+  Free     0   
10839     4.5     114  Varies with device       1,000+  Free     0   
10840     4.5  398307                 19M  10,000,000+  Free     0   

      Content Rating             Genres        Last Updated  \
10831       Everyone  Maps & Navigation       June 13, 2018   
10832       Everyone            Weather   February 16, 2014   
10833       Everyone  Books & Reference      March 23, 2014   
10834       Everyone          Education       June 18, 2017   
10835       Everyone           Business  September 29, 2016   
10836       Everyone          Education       July 25, 2017   
10837       Everyone          Education        July 6, 2018   
10838       Everyone            Medical    January 20, 2017   
10839     Mature 17+  Books & Reference    January 19, 2015   
10840       Everyone          Lifestyle       July 25, 2018   

              Current Ver         Android Ver  
10831           2.0.148.0          4.0 and up  
10832                 6.0          2.1 and up  
10833                 0.8          2.2 and up  
10834               1.0.0          4.1 and up  
10835               1.1.5          4.0 and up  
10836                1.48          4.1 and up  
10837                 1.0          4.1 and up  
10838                 1.0          2.2 and up  
10839  Varies with device  Varies with device  
10840  Varies with device  Varies with device

In [6]:
#Displaying the first 10 rows
ratings_df.head(10)

App                                  Translated_Review  \
0  10 Best Foods for You  I like eat delicious food. That's I'm cooking ...   
1  10 Best Foods for You    This help eating healthy exercise regular basis   
2  10 Best Foods for You                                                NaN   
3  10 Best Foods for You         Works great especially going grocery store   
4  10 Best Foods for You                                       Best idea us   
5  10 Best Foods for You                                           Best way   
6  10 Best Foods for You                                            Amazing   
7  10 Best Foods for You                                                NaN   
8  10 Best Foods for You                               Looking forward app,   
9  10 Best Foods for You              It helpful site ! It help foods get !   

  Sentiment  Sentiment_Polarity  Sentiment_Subjectivity  
0  Positive                1.00                0.533333  
1  Positive                0.25                0.288462  
2       NaN                 NaN                     NaN  
3  Positive                0.40                0.875000  
4  Positive                1.00                0.300000  
5  Positive                1.00                0.300000  
6  Positive                0.60                0.900000  
7       NaN                 NaN                     NaN  
8   Neutral                0.00                0.000000  
9   Neutral                0.00                0.000000

In [7]:
#Displaying the bottom 10 rows
ratings_df.tail(10)

App Translated_Review Sentiment  \
64285  Houzz Interior Design Ideas               NaN       NaN   
64286  Houzz Interior Design Ideas               NaN       NaN   
64287  Houzz Interior Design Ideas               NaN       NaN   
64288  Houzz Interior Design Ideas               NaN       NaN   
64289  Houzz Interior Design Ideas               NaN       NaN   
64290  Houzz Interior Design Ideas               NaN       NaN   
64291  Houzz Interior Design Ideas               NaN       NaN   
64292  Houzz Interior Design Ideas               NaN       NaN   
64293  Houzz Interior Design Ideas               NaN       NaN   
64294  Houzz Interior Design Ideas               NaN       NaN   

       Sentiment_Polarity  Sentiment_Subjectivity  
64285                 NaN                     NaN  
64286                 NaN                     NaN  
64287                 NaN                     NaN  
64288                 NaN                     NaN  
64289                 NaN                     NaN  
64290                 NaN                     NaN  
64291                 NaN                     NaN  
64292                 NaN                     NaN  
64293                 NaN                     NaN  
64294                 NaN                     NaN

In [8]:
# Checking the data type
store_df.dtypes

App                object
Category           object
Rating            float64
Reviews            object
Size               object
Installs           object
Type               object
Price              object
Content Rating     object
Genres             object
Last Updated       object
Current Ver        object
Android Ver        object
dtype: object

In [9]:
# Checking the data type
ratings_df.dtypes

App                        object
Translated_Review          object
Sentiment                  object
Sentiment_Polarity        float64
Sentiment_Subjectivity    float64
dtype: object

In [10]:
# Checking the number ofrows andd column
store_df.shape

(10841, 13)

In [11]:
# Checking the number ofrows andd column
ratings_df.shape

(64295, 5)

# **By diagnosing the Play Store DF & User Ratings DF, we get to know that:**

### IN Play Store DF
1.   There are 13 columns of properties with 10841 rows of data.
2.   Column 'Reviews', 'Size', 'Installs' and 'Price' are in the type of 'object'.
3.   Values of column 'Size' are strings representing size in 'M' as Megabytes, 'k' as kilobytes and also 'Varies with devices'.
4.   Values of column 'Installs' are strings representing install amount with symbols such as ',' and '+'.
5.   Values of column 'Price' are strings representing price with symbol '$'.
6.   There is only one float64 type column which is 'Rating'.

### IN User Ratings DF

1.  There are 5 columns of properties with 64295 rows of data.
2.  Column App,Translated_Review,Sentiment are in the type of 'object'.
3.  Column Sentiment_Polarity,Sentiment_Subjectivity are in the type of float64

## **THIRD STEP:-**
## **DATA CLEANING**

In data cleaning we will check for:


1.   Null values/NaN values
2.   Duplicates 
3.   Outlier Detection



In [12]:
# 1. CLEAN THE REVIEWS DATA AND CHANGE THE TYPE FROM 'OBJECT' TO 'FLOAT'

reviews = [i for i in store_df['Reviews']]

def clean_ratings(rating_list):

  """
  An 'M has been found while changing the type,
  this function will replace the 'M' with million
  """

  cleaned_data = []
  for review in rating_list:

    if 'M' in review:
      review = review.replace('M', '')
      review = float(review) * 1000000    # 1 M = 10,00,000
    cleaned_data.append(review)
    
  return cleaned_data

In [13]:
store_df['Reviews'] = clean_ratings(reviews)
store_df['Reviews'] = store_df['Reviews'].astype(float)

In [14]:
# FOUND A ROW WITH 1000+ IN SIZE COLUMN
# REMOVING IT AS WE ARE UNCERTAIN ABOUT IT

index = store_df[store_df['Size'] == '1,000+'].index
store_df.drop(axis = 0, inplace = True, index = index)

In [15]:
# 2. CLEAN SIZE DATA COLUMN AND CHANGE ITS TYPE FROM 'OBJECT' TO 'FLOAT'

sizes = [i for i in store_df['Size']]

def clean_size(size_list):

  """
  As sizes are repesented in 'M' or 'K', we
      remove 'M' and convert 'K' into 'M'
  """

  cleaned_data = []
  for size in size_list:

    if 'M' in size:
      size = size.replace('M', '')
      size = float(size)

    elif 'k' in size:
      size = size.replace('k', '')
      size = float(size)
      size = size / 1024    # 1 MB = 1024 KB

    else:
      size = float(0)       # REPRESENTING 'VARIES WITH DEVICE' WITH VALUE 0
    cleaned_data.append(size)
    
  return cleaned_data

In [16]:
store_df['Size'] = clean_size(sizes)
store_df['Size'] = store_df['Size'].astype(float)

In [17]:
store_df.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

    Reviews  Size     Installs  Type Price Content Rating  \
0     159.0  19.0      10,000+  Free     0       Everyone   
1     967.0  14.0     500,000+  Free     0       Everyone   
2   87510.0   8.7   5,000,000+  Free     0       Everyone   
3  215644.0  25.0  50,000,000+  Free     0           Teen   
4     967.0   2.8     100,000+  Free     0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

In [18]:
store_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10840 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10840 non-null  object 
 1   Category        10840 non-null  object 
 2   Rating          9366 non-null   float64
 3   Reviews         10840 non-null  float64
 4   Size            10840 non-null  float64
 5   Installs        10840 non-null  object 
 6   Type            10839 non-null  object 
 7   Price           10840 non-null  object 
 8   Content Rating  10840 non-null  object 
 9   Genres          10840 non-null  object 
 10  Last Updated    10840 non-null  object 
 11  Current Ver     10832 non-null  object 
 12  Android Ver     10838 non-null  object 
dtypes: float64(3), object(10)
memory usage: 1.2+ MB


In [19]:
# 3. CLEAN THE INSTALLS DATA AND CHANGE THE TYPE FROM 'OBJECT' TO 'FLOAT'

installs = [i for i in store_df['Installs']]

def clean_install(install_list):
  """
  This function removes the + and , from the data.
  There is one row where installs is 'Free'. Since there is only one,
  so, we make it 0.
  """
  
  clean_data = []
  for install in install_list:

    if '+' in install:
      install = install.replace('+', '')  # REMOVE +

    if ',' in install:
      install = install.replace(',', '')  # REMOVE ,
    
    install = int(install)
    clean_data.append(install)

  return clean_data

In [20]:
store_df['Installs'] = clean_install(installs)
store_df['Installs'] = store_df['Installs'].astype(float)

In [21]:
# 4. CLEAN THE PRICE DATA AND CHANGE THE TYPE FROM 'OBJECT' TO 'FLOAT'

prices = [i for i in store_df['Price']]

def clean_prices(price_list):
  """
  This function removes the '$' sign from the data.
  There is one row of price column with data as everyone.
  Making it free or 0.
  """
  clean_data = []
  for price in price_list:

    if '$' in price:
      price = price.replace('$', '')

    clean_data.append(price)
  
  return clean_data

In [22]:
store_df['Price'] = clean_prices(prices)
store_df['Price'] = store_df['Price'].astype(float)

In [23]:
store_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10840 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10840 non-null  object 
 1   Category        10840 non-null  object 
 2   Rating          9366 non-null   float64
 3   Reviews         10840 non-null  float64
 4   Size            10840 non-null  float64
 5   Installs        10840 non-null  float64
 6   Type            10839 non-null  object 
 7   Price           10840 non-null  float64
 8   Content Rating  10840 non-null  object 
 9   Genres          10840 non-null  object 
 10  Last Updated    10840 non-null  object 
 11  Current Ver     10832 non-null  object 
 12  Android Ver     10838 non-null  object 
dtypes: float64(5), object(8)
memory usage: 1.2+ MB


In [24]:
# CHECKING FOR NULL VALUES

store_df.isna().sum()

App                  0
Category             0
Rating            1474
Reviews              0
Size                 0
Installs             0
Type                 1
Price                0
Content Rating       0
Genres               0
Last Updated         0
Current Ver          8
Android Ver          2
dtype: int64

So, there are 1474 null values under 'Rating' column. So, we are going to replace with median of overall rating values.

In [25]:
def replace_with_median(series):
  """
  Takes a series and replace the null values with median
  """
  return series.fillna(series.fillna(series.median()))

In [26]:
store_df['Rating'] = store_df['Rating'].transform(replace_with_median)
store_df['Rating'] = store_df['Rating'].astype(float)

In [27]:
# REMOVE THE RECORD WHERE 'TYPE' IS HAVING NULL VALUE

index = store_df[store_df['Type'].isna()].index
store_df.drop(axis = 0, inplace = True, index = index)

In [28]:
# CHECKING FOR NULL VALUES AGAIN

store_df.isna().sum()

App               0
Category          0
Rating            0
Reviews           0
Size              0
Installs          0
Type              0
Price             0
Content Rating    0
Genres            0
Last Updated      0
Current Ver       8
Android Ver       2
dtype: int64

In [29]:
# FILLING THE NULL VALUES OF CURRENT VERSION WITH 1.0

store_df['Current Ver'] = store_df['Current Ver'].fillna(1.0)

In [30]:
# FILLING THE NULL VALUES OF ANDROID VERSION WITH PREVIOUS DATA

store_df['Android Ver'] = store_df['Android Ver'].fillna(method = 'pad')

In [31]:
# CHECKING FOR NULL VALUES AGAIN

store_df.isna().sum()

App               0
Category          0
Rating            0
Reviews           0
Size              0
Installs          0
Type              0
Price             0
Content Rating    0
Genres            0
Last Updated      0
Current Ver       0
Android Ver       0
dtype: int64

In [32]:
# DROPPING THE DUPLICATE ROWS BASED ON APPS

store_df.drop_duplicates(subset = ['App'], inplace=True)

In [33]:
# CHECK ON STATISTICAL INFORMATION OF THE DATAFRAME

store_df.describe()

Rating       Reviews         Size      Installs        Price
count  9658.000000  9.658000e+03  9658.000000  9.658000e+03  9658.000000
mean      4.192431  2.166150e+05    17.805940  7.778312e+06     1.099413
std       0.496422  1.831413e+06    21.495867  5.376100e+07    16.853021
min       1.000000  0.000000e+00     0.000000  0.000000e+00     0.000000
25%       4.000000  2.500000e+01     2.900000  1.000000e+03     0.000000
50%       4.300000  9.670000e+02     9.100000  1.000000e+05     0.000000
75%       4.500000  2.940800e+04    25.000000  1.000000e+06     0.000000
max       5.000000  7.815831e+07   100.000000  1.000000e+09   400.000000

*  Half of the apps are less than 10 MB.
*  Maximum size of the available app in this data is 100 MB.

In [34]:
# CHECKING FOR THE USER REVIEW DATAFRAME

review_df = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/AlmaBetter/PROJECTS/EDA/PLAY STORE/User Reviews.csv')
review_df.head(10)

App                                  Translated_Review  \
0  10 Best Foods for You  I like eat delicious food. That's I'm cooking ...   
1  10 Best Foods for You    This help eating healthy exercise regular basis   
2  10 Best Foods for You                                                NaN   
3  10 Best Foods for You         Works great especially going grocery store   
4  10 Best Foods for You                                       Best idea us   
5  10 Best Foods for You                                           Best way   
6  10 Best Foods for You                                            Amazing   
7  10 Best Foods for You                                                NaN   
8  10 Best Foods for You                               Looking forward app,   
9  10 Best Foods for You              It helpful site ! It help foods get !   

  Sentiment  Sentiment_Polarity  Sentiment_Subjectivity  
0  Positive                1.00                0.533333  
1  Positive                0.25                0.288462  
2       NaN                 NaN                     NaN  
3  Positive                0.40                0.875000  
4  Positive                1.00                0.300000  
5  Positive                1.00                0.300000  
6  Positive                0.60                0.900000  
7       NaN                 NaN                     NaN  
8   Neutral                0.00                0.000000  
9   Neutral                0.00                0.000000

In [35]:
# CHECKING FOR DATA TYPES

review_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 64295 entries, 0 to 64294
Data columns (total 5 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   App                     64295 non-null  object 
 1   Translated_Review       37427 non-null  object 
 2   Sentiment               37432 non-null  object 
 3   Sentiment_Polarity      37432 non-null  float64
 4   Sentiment_Subjectivity  37432 non-null  float64
dtypes: float64(2), object(3)
memory usage: 2.5+ MB


In [36]:
# CHECKING FOR THE NULL VALUES

review_df.isna().sum()

App                           0
Translated_Review         26868
Sentiment                 26863
Sentiment_Polarity        26863
Sentiment_Subjectivity    26863
dtype: int64

In [37]:
# REPLACING THE NULL VALUES OF TRANSLATED REVIEW WITH ITS PREVIOUS ONE

review_df['Translated_Review'] = review_df['Translated_Review'].fillna(method = 'pad')

In [38]:
# REPLACING THE NULL VALUES OF SENTIMENT WITH ITS PREVIOUS ONE

review_df['Sentiment'] = review_df['Sentiment'].fillna(method = 'pad')

In [39]:
# REPLACING THE NULL VALUES OF SENTIMENT POLARITY AND SENTIMENT SUBJECTIVITY WITH ITS MEDIAN

review_df['Sentiment_Polarity'] = review_df['Sentiment_Polarity'].transform(replace_with_median)
review_df['Sentiment_Subjectivity'] = review_df['Sentiment_Subjectivity'].transform(replace_with_median)

In [40]:
# CHECKING FOR THE NULL VALUES AGAIN

review_df.isna().sum()

App                       0
Translated_Review         0
Sentiment                 0
Sentiment_Polarity        0
Sentiment_Subjectivity    0
dtype: int64

#Exaploratory Analysis and Visualization